contracts/src/Teleporter/TeleporterMessenger.sol

```diff
@@ -745,19 +754,32 @@ contract TeleporterMessenger is ITeleporterMessenger, ReentrancyGuards {
      */
     function _storeFailedMessageExecution(
         bytes32 originBlockchainID,
+        bytes32 messageID,
         TeleporterMessage memory message
     ) private {
-        receivedFailedMessageHashes[originBlockchainID][message.messageID] =
-            keccak256(abi.encode(message));
+        receivedFailedMessageHashes[messageID] = keccak256(abi.encode(message));
 
         // Emit a failed execution event for anyone monitoring unsuccessful messages to retry.
-        emit MessageExecutionFailed(originBlockchainID, message.messageID, message);
+        emit MessageExecutionFailed(originBlockchainID, messageID, message);
     }
 
     /**
-     * @dev Returns the next message ID to be used to send a message to the given `originBlockchainID`.
+     * @dev Calculates the message ID for the given source blockchain ID and message nonce.
      */
-    function _getNextMessageID(bytes32 originBlockchainID) private view returns (uint256) {
-        return latestMessageIDs[originBlockchainID] + 1;
+    function _calculateMessageID(
+        bytes32 sourceBlockchainID,
+        uint256 nonce
+    ) private view returns (bytes32) {
+        return sha256(abi.encode(address(this), sourceBlockchainID, nonce));
```

Comment:
why do we opt for sha256 vs keccak256

Reply:
A tiny bit of research seems to say keccak is both cheaper (marginally) and more secure (not so important here, but nice to get for free).

Reply:
Agreed, was already use keccak256 else in the contract also, so I switched to it as a part of #209 to be consistent.